### DISCUSSION TITLE
[help wanted] How to keep extmark from removed after commenting code

### DISCUSSION THREAD

**Discussion by shellRaining** [nvim-mini/mini.nvim#959] [Q&A] [answered]
hello, very thanks for your nice work! `mini.indentscope` has cool effect, and provide many inspiration when I wrote my plugin. I am writing a indent (chunk) plugin, and encounter a bug, when comment a line which has extmark, the extmark was remove, detail can be seen here [https://github.com/shellRaining/hlchunk.nvim/issues/109](https://github.com/shellRaining/hlchunk.nvim/issues/109)

I found that `mini.indentscope` also has lazy feat and can keep extmark after commenting a line, But I feel very confused about how he achieved this effect.

I originally wanted to post on Reddit, but it seems that my account and IP have been banned. I'm sorry if this has caused any inconvenience.

**Comment by echasnovski** [comment, marked as answer]
Hi, @shellRaining! I remember seeing screenshots of that chunk highlighting and thinking that it was beautiful. Nice work!

Applying `nvim_buf_set_lines()` indeed "squashes" all extmarks to be on either side of set lines. This is essentially close to manually deleting those lines (which moves extmarks) and pasting text (which also might move extmarks).

Extmarks from 'mini.indentscope' are not exceptions. The way 'mini.indentscope' **makes it look** like extmarks don't move is by [redrawing on every 'TextChanged' event](https://github.com/echasnovski/mini.nvim/blob/d7e05b30d842e543366ff5474076a0726f783a32/lua/mini/indentscope.lua#L679) immediately before Neovim redraws the screen.

In turn, 'mini.indentscope' decides whether it should redraw scope immediately or not is by comparing currently drawn scope with target. If [those scopes intersect](https://github.com/echasnovski/mini.nvim/blob/d7e05b30d842e543366ff5474076a0726f783a32/lua/mini/indentscope.lua#L933-L938), then it draws with delay 0 the whole scope right away.

So in short: it re-adds extmarks very quickly after every text change.

------

For the future, this type of question is better be posted in [Discussions](https://github.com/echasnovski/mini.nvim/discussions) directly. I've transferred this one myself.

**Reply by shellRaining** [reply]
Thank you very much for your help. I understand how to solve it, and I noticed that you mentioned `lockmarks` in the official repositories of `Neovim` and `comment.nvim`. Can we use this to ensure that calling `nvim_buf_set_lines()` not clear extmark? 

I have searched a lot of information online, but they are all used in conjunction with `visualMode`, and I still don't quite understand how to use it. Thanks!

**Reply by echasnovski** [reply]
> ... I noticed that you mentioned `lockmarks` in the official repositories of `Neovim` and `comment.nvim`. Can we use this to ensure that calling `nvim_buf_set_lines()` not clear extmark?

Unfortunately, no. The `lockmarks` command modifier only preserves regular marks (like `<`/`>` for latest visual selection, etc.), not extended marks (which are used for all sorts of things).

The only way (right now) to preserve extmarks is to precisely use `nvim_buf_set_text()` to edit only those parts of lines which need to be changed. 

But (as always) it is not that simple: preserving extmarks is not a universally good thing. I spent quite a bit of time contemplating this in 'mini.comment' (and built-in commenting by extension). In the end, using `nvim_buf_set_lines()` and **not** preserving extmarks seems like the best thing to do for commenting. Here is a [full PhD size comment with motivation](https://github.com/echasnovski/mini.nvim/blob/d7e05b30d842e543366ff5474076a0726f783a32/lua/mini/comment.lua#L260-L287).

**Reply by shellRaining** [reply]
Thank you for your clarification. I saw that you mentioned a set of data. When commenting four thousand lines, using `nvim_buf_set_text()` will consume about 400ms. There are two points of doubt:

1. Will we often comment out so many lines of code? Are there other reasons why we choose `nvim_buf_set_lines()`?
2. Can we divide the four thousand lines of code into blocks (such as one win_height as a block), and then send them to the task queue to avoid lagging as much as possible?

**Reply by echasnovski** [reply]
> 1. Are there other reasons why we choose `nvim_buf_set_lines()`?

Yes, [see the next point](https://github.com/echasnovski/mini.nvim/blob/d7e05b30d842e543366ff5474076a0726f783a32/lua/mini/comment.lua#L273-L282).

> 2\. Can we divide the four thousand lines of code into blocks (such as one win_height as a block), and then send them to the task queue to avoid lagging as much as possible?

Sure, but that is a more complex approach and won't solve other issues with preserving extmarks.

**Comment by CoataocCreate** [comment]
To prevent `extmark` from being removed after commenting code in Vim, you'll need to ensure that Vim doesn't strip out such marks when commenting or uncommenting lines. Here’s a step-by-step approach to achieve this:

### Understanding Extmarks in Vim

In Vim, `extmark` (external mark) is a feature that allows plugins and scripts to place markers in a buffer that persist even when the text is edited or modified. However, when lines are commented or uncommented, Vim typically removes or adjusts these marks by default, which can interfere with certain plugin functionalities.

### Solution: Use `preserve` Option

Vim provides an option called `preserve`, which instructs Vim to preserve extmarks when performing certain operations like commenting and uncommenting lines. You can set this option to maintain your extmarks intact.

1. **Add `preserve` Option in Vimrc:**

   Open your Vim configuration file (`~/.vimrc` or `~/.config/nvim/init.vim` for Neovim) and add the following line:

   ```vim
   set preserve+=,n
   ```

   - `+=` appends to the existing value of `preserve`.
   - `n` specifies that the extmarks (`m` and `M`) should be preserved.

   This setting ensures that extmarks are not disturbed when commenting or uncommenting lines.

2. **Verify and Apply:**

   After adding the above line to your Vim configuration, save the file and restart Vim or reload the configuration (`:source ~/.vimrc` or `:source ~/.config/nvim/init.vim` in Neovim).

3. **Test Extmarks:**

   Test whether extmarks are preserved when commenting or uncommenting lines in Vim. Place some extmarks using a plugin or script, comment out lines containing these marks, and verify that the marks are still present and functional.

### Example Usage

Assuming you're using a plugin that utilizes extmarks (e.g., for highlighting purposes), here's a hypothetical example of how you might use extmarks:

```vim
" Place an extmark at current cursor position
:call setpos('.', [0, line('.'), col('.'), 0])

" Comment out current line using a plugin (e.g., commentary.vim)
:Commentary

" Ensure the extmark is preserved and still functional
```

### Conclusion

By setting `preserve` in your Vim configuration, you ensure that extmarks are retained even after performing operations like commenting or uncommenting lines. This approach helps maintain the functionality provided by plugins or scripts that rely on extmarks for various tasks such as highlighting, navigation, or custom operations within Vim.
